File: LeetCode/150/solution.java
// 150. Evaluate Reverse Polish Notation
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> st = new Stack<>();
        for(String token: tokens){
            if(token.length() == 1 && token.charAt(0) < 48){
                int top1 = st.pop();
                int top2 = st.pop();
                int val = resolve(top2, top1, token.charAt(0));
                st.push(val);
            }
            else st.push(Integer.parseInt(token));
        }
        return st.pop();
    }

    private int resolve(int x, int y, char op){
        if(op=='+') return x + y;
        else if(op=='-') return x - y;
        else if(op=='*') return x * y;
        else return x / y;
    }
}

File: LeetCode/150/solution.cpp
// 150. Evaluate Reverse Polish Notation
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        int sz = tokens.size();
        for(int i=0; i<sz; i++){
            if(isOperator(tokens[i])){
                int top1 = st.top();
                st.pop();
                int top2 = st.top();
                st.pop();
                int val = operation(top2, top1, tokens[i]);
                st.push(val);
            }
            else st.push(stoi(tokens[i]));
        }
        return st.top();
    }

    bool isOperator(string s){
        return (s=="+" || s=="-" || s=="*" || s=="/");
    }

    int operation(int x, int y, string op){
        if(op=="+") return x + y;
        if(op=="-") return x - y;
        if(op=="*") return x * y;
        if(op=="/") return (int)floor(x / y);
        return 0;
    }
};